### PR TITLE
feat: upstream Solmate's bound 

### LINKED ISSUE #48: feat: upstream Solmate's `bound`
This method is useful for wrapping inputs of a fuzz test into a certain range, and t11s has given approval to upstream this. Scope:
- Add [`bound`](https://github.com/Rari-Capital/solmate/blob/9f16db2144cc9a7e2ffc5588d4bf0b66784283bd/src/test/utils/DSTestPlus.sol#L115-L136)
- Add the associates [tests](https://github.com/Rari-Capital/solmate/blob/main/src/test/DSTestPlus.t.sol)
    - Add `bound` tests that cover `type(uint256).max` as the upper bound input, since this is an edge case handled separately within `bound`
    - Change the `testFail` tests to the `expectRevert` style

I don't plan on working on this over the next few days, so this is open for anyone to take 🙂

### PR DESCRIPTION
**Pull request by sscobes**
Added ```bound``` from Solmate and added the associated tests from the repo as well. All ```testFail``` style tests were updated to ```expectRevert``` and an additional test for ```type(uint256).max``` as the upper bound was also added to test edge case.

Closes #48

### CODE REVIEW

File: src/Test.sol [hunk 1]
```diff
@@ -322,6 +345,25 @@ abstract contract Test is DSTest {
...
+
+        if (percentDelta > maxPercentDelta) {
+            emit log("Error: a ~= b not satisfied [uint]");
+            emit log_named_uint("    Expected", b);
+            emit log_named_uint("      Actual", a);
+            emit log_named_decimal_uint(" Max % Delta", maxPercentDelta, 18);
+            emit log_named_decimal_uint("     % Delta", percentDelta, 18);
+            fail();
+        }
+    }
```
Comment (ZeroEkkusu) [inline src/Test.sol:366]: We don't need this for `bound`.

File: src/test/StdCheats.t.sol [hunk 2]
```diff
@@ -76,6 +76,60 @@ contract StdCheatsTest is Test {
...
+
+    function testRevertBoundMinBiggerThanMax() public {
+        vm.expectRevert(bytes("MAX_LESS_THAN_MIN"));
+        bound(5, 100, 10);
+    }
+
+    function testRelApproxEqBothZeroesPasses() public {
+        assertRelApproxEq(0, 0, 1e18);
+        assertRelApproxEq(0, 0, 0);
+    }
```
Comment (ZeroEkkusu) [inline src/test/StdCheats.t.sol:97]: We don't need this for `bound`.

File: src/test/StdCheats.t.sol [hunk 3]
```diff
@@ -76,6 +76,60 @@ contract StdCheatsTest is Test {
...
+    ) public {
+        if (min > max) (min, max) = (max, min);
+
+        uint256 bounded = bound(num, min, max);
+
+        assertGe(bounded, min);
+        assertLe(bounded, max);
+    }
+
+    function testBound_Uint256Max() public {
```
Comment (ZeroEkkusu) [inline src/test/StdCheats.t.sol:112]: Let's keep the code style consistent; use `PascalCase` here.
```suggestion
    function testBoundUint256Max() public {
```

File: src/test/StdCheats.t.sol [hunk 4]
```diff
@@ -76,6 +76,60 @@ contract StdCheatsTest is Test {
...
+        uint256 min,
+        uint256 max
+    ) public {
+        if (max == min) {
+            unchecked {
+                min++; // Overflow is handled below.
+            }
+        }
+
+        if (max > min) (min, max) = (max, min);
```
Comment (ZeroEkkusu) [inline src/test/StdCheats.t.sol:127]: Use `vm.assume` here instead.
```suggestion
        vm.assume(min > max);
```

**Review by ZeroEkkusu — CHANGES_REQUESTED** [review]
Thanks!

We need to make a couple of changes. Left comments for those.

File: src/test/StdCheats.t.sol (hunk 4 above)
Comment (sscobes) [reply src/test/StdCheats.t.sol]: Updated to use ```vm.assume```.

File: src/test/StdCheats.t.sol (hunk 3 above)
Comment (sscobes) [reply src/test/StdCheats.t.sol]: Updated function naming style.

File: src/test/StdCheats.t.sol (hunk 2 above)
Comment (sscobes) [reply src/test/StdCheats.t.sol]: Removed  from StdCheats.t.sol.

File: src/Test.sol (hunk 1 above)
Comment (sscobes) [reply src/Test.sol]: Removed  from StdCheats.t.sol.

**Comment by sscobes** [comment]
Made requested changes in commit 580e4aca4166247eb29830fd014a3d4e97bb7063.

**Review by ZeroEkkusu — CHANGES_REQUESTED** [review]
Nice. I want to check with @transmissions11 if this is the intended behavior of `bound`, because I expected it to work differently.

The main issue I have is that `min` is sometimes not included when `max` is `type(uint256).max`. Is this the intended behavior?

The code is from Solmate, btw.

### Examples

*x --> result*

#### `bound(x, 40, 50);`

0 --> 44
1 --> 45
2 --> 46
3 --> 47
4 --> 48

Notes:
- starts from `44`

#### `bound(x, 0, type(uint256).max);`

0 --> 0
1 --> 0
2 --> 2
3 --> 3
4 --> 4

Notes:
- skips `1`
- repeats at the beginning

#### `bound(x, type(uint256).max - 1, type(uint256).max);`

0 --> 1157...35
1 --> 1157...35
2 --> 1157...35
3 --> 1157...35
4 --> 1157...35

Notes:
- `min` is not included

#### `bound(x, type(uint256).max - 6, type(uint256).max);`

0 --> 1157...33
1 --> 1157...33
2 --> 1157...34
3 --> 1157...35
4 --> 1157...30

Notes:
- `min` is not included
- repeats at the beginning
- starts from `1157...33`

### Alternative

If this indeed is not the intended behavior, let's come up with a better algo:

```solidity
function bound(
    uint256 x,
    uint256 min,
    uint256 max
) internal returns (uint256) {
    require(min <= max, "MAX_LESS_THAN_MIN");

    uint256 size = max - min;

    if (size == 0) {
        return min;
    }

    if (size == type(uint256).max) {
        return x;
    }

    ++size; // make `max` inclusive
    uint256 mod = x % size;
    return min + mod;
}
```

It can perfectly bound the input every time:

0 --> `min`
1 --> `min` + 1
.
.
n --> `max`
n + 1 --> `min`
n + 2 --> `min` + 1
*(loops)*

#### Examples

##### `bound(x, type(uint256).max - 1, type(uint256).max);`

0 --> 1157...34
1 --> 1157...35
2 --> 1157...34
3 --> 1157...35
4 --> 1157...34

Notes:
- `min` is included

##### `bound(x, type(uint256).max - 6, type(uint256).max);`

0 --> 1157...29
1 --> 1157...30
2 --> 1157...31
3 --> 1157...32
4 --> 1157...33

Notes:
- `min` is included
- does not repeat at the beginning
- starts from `1157...29`

**Comment by transmissions11** [comment]
yea that bound impl is kinda weird, i stole it from stackoverflow and modified it a bit so not confident in all its properties. lets go with your impl— looks great! have you run it thru the tests in solmate? https://github.com/Rari-Capital/solmate/blob/c8278b3cb948cffda3f1de5a401858035f262060/src/test/DSTestPlus.t.sol#L25-L52

**Comment by ZeroEkkusu** [comment]
@transmissions11, yea, i ran it thru the tests from solmate and my own tests; it works well. want me to open a pr to solmate?

@sscobes, i just pushed the new implementation to this branch. can you modify the tests to work with it? additionally, let's change the names of the revert tests to `testCannotBoundMaxLessThanMin`.

**Comment by transmissions11** [comment]
> @transmissions11, yea, i ran it thru the tests from solmate and my own tests; it works well. want me to open a pr to solmate?

awesome, yes pls!

**Comment by sscobes** [comment]
Updated tests to work with the new bound implementation and changed the name for the mentioned test functions @ZeroEkkusu

File: src/test/StdCheats.t.sol [hunk 5]
```diff
@@ -76,6 +76,48 @@ contract StdCheatsTest is Test {
...
+        if (min > max) (min, max) = (max, min);
+
+        uint256 bounded = bound(num, min, max);
+
+        assertGe(bounded, min);
+        assertLe(bounded, max);
+    }
+
+    function testBoundUint256Max() public {
+        assertEq(bound(0, type(uint256).max - 1, type(uint256).max), type(uint256).max - 1);
```
Comment (ZeroEkkusu) [inline src/test/StdCheats.t.sol:108]: We should also to test the case when the `result` is `type(uint256).max`.
```suggestion
        assertEq(bound(0, type(uint256).max - 1, type(uint256).max), type(uint256).max - 1);
        assertEq(bound(1, type(uint256).max - 1, type(uint256).max), type(uint256).max);
```

**Review by ZeroEkkusu — CHANGES_REQUESTED** [review]
Just this one nit. Otherwise, LGTM.

File: src/test/StdCheats.t.sol (hunk 5 above)
Comment (sscobes) [reply src/test/StdCheats.t.sol:108]: done :)

**Review by ZeroEkkusu — APPROVED** [review]

**Review by mds1 — APPROVED** [review]
This looks really great! Thanks a lot @sscobes for getting this done, and @ZeroEkkusu great fix on the bound implementation
